Experiment: Add lightbox to Image block using directives

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This is part of the experiments section with the interactivity API in revising core blocks.
It uses directives from the [Directives hydration system](https://github.com/WordPress/block-interactivity-experiments/issues/80)

## Why?
It would enable native support for image lightboxes, a commonly used feature in many websites, without the use of jQuery or other external libraries.

## How?
It adds directives on click to images, presenting a larger version of the image along with an overlay, which can be hidden with an additional click, by scrolling, or pressing the escape key.

## Testing Instructions
1. Add an image to a post
2. Test enabling and disabling the lightbox using image's Advanced panel
3. View the post
4. Click on the image

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
1. Tab to the image.
2. Press Enter to open the lightbox.
3. Test Tab or Shift + Tab — the lightbox should close and focus should go back to the image.
4. Also test pressing Escape to close the lightbox.

## Notes
- This is an early implementation; the idea is for themes to be able to enable or disable the lightbox behavior, for child themes to be able to override that behavior, and for users to be able to toggle lightboxes for all images sitewide or at the block level.
- This is part of a broader initiative to enable adding Interactivity API behaviors to components using the editor.
- Currently, we can just toggle the lightbox for a particular image at the block level under the Advanced panel (see below)
- The CSS implementation is not finalized. Ideally, I'd like to calculate the screen size and set the image to say 80% of the screen width, which will likely require some calculation outside of CSS. Is there a best practice for doing this?
- I'm looking for feedback on the design and overall approach

## Screenshots or screencast <!-- if applicable -->

### Editor

https://user-images.githubusercontent.com/5360536/232110768-4749230c-facd-4d3e-b7b4-c46eaddf8b6d.mp4

### Frontend

https://user-images.githubusercontent.com/5360536/230152865-5957d4b0-dd71-41ad-b54d-0f1e4490c991.mp4

artemiomorales committed:
Add lightbox to image block

First pass at adding lightbox.

Note: Added custom implementation of Preact Portal
because the children[0] declaration in the render method
was erroneous and undefined, preventing it from working
as expected.

artemiomorales committed:
Add logic for hiding lightbox on esc key press and overlay click

artemiomorales committed:
Improve styles and add note to add conditional for lightbox markup

artemiomorales committed:
Add editor UI and attribute for toggling lightbox

jasmussen commented:
This is really cool. Here's a GIF of the current state of the branch, showing the lightbox toggle in advanced, and how that looks when clicked on the frontend:

![lightbox](https://user-images.githubusercontent.com/1204802/231994567-a6b4fa8a-9c1f-41b7-b13a-d2acf3da553a.gif)

Amazing work. Just works.

A couple of notes:

* We probably want this toggle to be a block support, and sit in the styles tab of the image block inspector. Would not be a blocker for an initial version, but just thinking in terms of next steps.
* It should probably be a toggle control instead of a checkbox. I think of toggles as equivalents to light switches: immediate effect, whereas a checkbox is usually paired with an "Apply" button.
* The white scrim looks good. Outside of full-customization of this scrim color, could you default to using the site background color?
* Can you make the scrim fade in and out as you open/close?
* Speaking of, it's likely going to be the most useful as a global styles toggle. I wonder if for a first version, it could live in Global Styles > Blocks > Image _only_? That might let us get this out there, as adjacent features get hashed out.
* I like that when you scroll, it "zooms out". I like that you used the zoom and zoom out cursors. I like that I can still right-click/open the source image in a new tab.

Note that if I also link the image to a destination, if I click the link, the lightbox invokes quickly, and then shortly after, the link is navigated to. We should probably just disable the lightbox entirely when an image is linked. This is likely worth doing also for the checkbox itself, or the design tool when we get a chance to upgrade it: it could be grayed out and have help-text that says something like "The lightbox behavior is disabled for linked images."

There's occasionally a "blink", or a flash of the image itself, as if it grows fully white for a nanosecond. It's hard to reproduce:

![blink](https://user-images.githubusercontent.com/1204802/231997468-5dc91b6f-3574-42e3-9357-63afcddd09ae.gif)

When you click a small image that's a ways down on the page, it opens in place:
<img width="2414" alt="Screenshot 2023-04-14 at 11 03 04" src="https://user-images.githubusercontent.com/1204802/231998046-eaee8147-e2f6-47d9-8b77-b78e8d5e338b.png">

<img width="2414" alt="Screenshot 2023-04-14 at 11 03 07" src="https://user-images.githubusercontent.com/1204802/231998066-f5ba67a7-1a3b-4d9f-b400-eefa05b80c83.png">

As shown in the example above, the image could be both bigger, and ideally centered in the currently visible viewport. Something like this:
<img width="2032" alt="Screenshot 2023-04-14 at 11 09 41" src="https://user-images.githubusercontent.com/1204802/232000384-538d72c6-a1ad-4d42-825a-ebe07f02d3d8.png">

That's where it gets a little bit tricky, and will likely require some difficult JS to get just right. In the above, just for test purposes, I simply applied the following CSS to the image:

```
    cursor: zoom-out;
    position: fixed;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 100%;
    object-fit: contain;
    padding: 2%;
```

That's not going to work too well, because it can't be animated in a performant way. But the principle is the same — show the image as big as the viewport is, (but bounded by max-resolution), and ideally centered in the viewport. What do you think?

gziolo commented:
It will be essential to ensure that all the accessibility requirements are met. I added the `Need Accessibility Feedback` label. I found an article by Graham Armfield that [covers the specification for an accessible lightbox](https://www.coolfields.co.uk/2011/12/specification-for-an-accessible-lightbox/). I'm not sure if that is the best resource, but the list shared sounds reasonable:

> I’ve tried to crystallise the accessibility requirements down into these few bullet points:
>
> 1. Links that trigger lightbox:
>   a. Must be able to get and show keyboard focus
>   b. Must announce that they open an image in a lightbox – for screen readers and for sighted keyboard users
> 2. When lightbox opens, focus must be placed somewhere sensible within the lightbox – suggest a Close link above or below picture.
> 3. Wherever focus is within the lightbox it must be visible.
> 4. The large image within the lightbox should have appropriate alt text added – especially if it is also a link to close the lightbox.
> 5. Either of these:
>   a. Tabbing beyond the last link in the lightbox or shift-tabbing before the first link should close the lightbox
>   b. Tabbing should cycle within the actionable links within the lightbox until user clicks a close link.
> 6. Whenever lightbox is closed, focus should be returned to whatever link/or image opened it in the first place
> 7. If javascript is not present in the browser the page should behave appropriately.

That should be a good starting point for the discussion. It's very likely that the link that triggers the lightbox doesn't need to be visible until someone uses the assistive technologies. In effect, it could use the [`screen-reader-text`](https://make.wordpress.org/accessibility/handbook/markup/the-css-class-screen-reader-text/) CSS class that exists in WordPress core.

The Navigation block can also be used as a reference because when it's displayed as a small menu icon, it opens an accessible overlay with the close button.

artemiomorales commented:
@jasmussen Great, thanks for the feedback!

> - We probably want this toggle to be a block support, and sit in the styles tab of the image block inspector. Would not be a blocker for an initial version, but just thinking in terms of next steps.

Ok got it. This will likely come along with some more work on behaviors that I or someone else may take on — we can keep this in mind as we devise those next steps.

> - It should probably be a toggle control instead of a checkbox. I think of toggles as equivalents to light switches: immediate effect, whereas a checkbox is usually paired with an "Apply" button.
> - The white scrim looks good. Outside of full-customization of this scrim color, could you default to using the site background color?
> - Can you make the scrim fade in and out as you open/close?
>
> Note that if I also link the image to a destination, if I click the link, the lightbox invokes quickly, and then shortly after, the link is navigated to. We should probably just disable the lightbox entirely when an image is linked. This is likely worth doing also for the checkbox itself, or the design tool when we get a chance to upgrade it: it could be grayed out and have help-text that says something like "The lightbox behavior is disabled for linked images."
>
> There's occasionally a "blink", or a flash of the image itself, as if it grows fully white for a nanosecond. It's hard to reproduce:

Thanks for the feedback! I can make these changes and address these issues 👍

> Speaking of, it's likely going to be the most useful as a global styles toggle. I wonder if for a first version, it could live in Global Styles > Blocks > Image only? That might let us get this out there, as adjacent features get hashed out.

This was something mentioned during a discussion with @luisherranz and @SantosGuillamot as well. I believe the end goal is to incorporate this into Global Styles just like you mention here, though a first pass would be to just allow users to enable the lightbox at the theme level with theme.json.

> As shown in the example above, the image could be both bigger, and ideally centered in the currently visible viewport. Something like this...
> That's where it gets a little bit tricky, and will likely require some difficult JS to get just right. In the above, just for test purposes, I simply applied the following CSS to the image:
>
>    cursor: zoom-out;
>    position: fixed;
>    top: 0;
>    right: 0;
>    bottom: 0;
>    left: 0;
>    width: 100%;
>    height: 100%;
>    object-fit: contain;
>    padding: 2%;
>
> That's not going to work too well, because it can't be animated in a performant way. But the principle is the same — show the image as big as the viewport is, (but bounded by max-resolution), and ideally centered in the viewport. What do you think?

I think this sounds great. A heads up that this lightbox behavior is based on the behavior used at Medium.com. I've set up a test article [here](https://medium.com/@artemio.morales/test-lightbox-f3ff616557ee), though you can also go to any Medium article and click on an image to see it. On our end, I believe I'll need to calculate the width of the screen to set the image scale.

One question I have: How do you propose we handle this behavior on mobile? While the Medium behavior works well on desktop, it's a bit wonky in mobile web:

## Mobile

https://user-images.githubusercontent.com/5360536/232109918-ae28dedb-105a-404d-aebd-4e5a12244554.mp4

## Desktop

https://user-images.githubusercontent.com/5360536/232109945-dfac2a48-8f88-4384-bfb4-3e5fc7ab0f69.mp4

joedolson commented:
A lightbox is basically just a modal dialog wearing a costume, and it should follow the same accessibility protocol that any modal needs to follow. If this was built on top of existing modal implementations, that would probably be a good plan.

jasmussen commented:
> Thanks for the feedback! I can make these changes and address these issues 👍

Before you make too many changes to where the toggle lives, would be good to check with @SaxonF, who I think had a design putting this in the link dialog. Furthermore, on reflection I'm not sure it's valuable as a style option after all, since it's a bit more behavioral in nature — definitely a judgement call. 

The main motivation to moving it to a block supports was to surface it in Global Styles, so that it could become a single "set it and forget it" toggle: _make all images open a lightbox unless linked_. Saxon what do you think? Are there use cases where you would want to disable this? Also CC: @jameskoster as you've thought about global settings like these. Is there a way to make it trivial for Artemio to move forward, but still in the future make it a global toggle? Maybe it can continue to live in "Advanced" for now?

jameskoster commented:
I can't get this working for some reason. I see the "Lightbox" checkbox in the Advanced panel, but toggling it on does nothing on the frontend. Tried adding/removing links in the link control – no dice 🤔 

> who I think had a design putting this in the link dialog

This seems sensible to me. Otherwise it's not really clear which behaviour would take precedence if you were to enable lightbox _and_ add a link (either to media file or attachment page). On that note, it may make sense to see this option _only_ when a link to the media file is added?

I don't have a strong feeling about the global implementation, but on the whole it feels like more of a contextual affordance, similar to making a link open in a new tab. I'd probably lean more towards a user preference for choosing the default behaviour when inserting images, rather than a global setting that gets unilaterally applied across the site.

jasmussen commented:
My main concern with adding it to the link dialog is that if I wanted to retroactivelly enable it on every image on my site with thousands of posts, wouldn't I have to edit them all?

jameskoster commented:
Yes, that's the trade-off. It feels like a separate thing to me though, and could be handled in a plugin.

If we _did_ make this a global style property, what would happen if an image is set to link to the attachment page, and lightbox is true?

jasmussen commented:
> If we did make this a global style property, what would happen if an image is set to link to the attachment page, and lightbox is true?

If an image has _any_ kind of clickable link, my take is the lightbox should just not fire ever. So I could see a single toggle "show lightbox for images" somewhere, and then if an image is linked, it just won't fire there.

jameskoster commented:
That sounds a bit confusing to me, especially as a 'style' property. 🤔

jasmussen commented:
Yeah, style may not be the best place for it. That said, lightbox for only unlinked images seems to be the industry standard as far as my experience goes. Have you seen otherwise?

jameskoster commented:
Some members of the design team spoke about this recently in a huddle. 

The global element gets quite tricky, and probably needs more design. Is it a site setting, if so where does that live? Is it part of global styles, and if so how do we make the interactions with link control feel intuitive? How do we enable local overrides at different levels (page / template / pattern / block)?

With all the said, the best course of action may be to design a really good UI for the most atomic scenario, IE a single image block, and extrapolate out from there when we've had more time to consider the global scope. What do y'all think?

beafialho commented:
Has it been considered, instead of a _zoom in_ motion, which in my opinion can be a bit too much, something like this?

https://user-images.githubusercontent.com/65220155/232794170-a4d13192-9ec7-495b-8459-d3ee13fd4234.mp4

https://user-images.githubusercontent.com/65220155/232794245-415ec3eb-72cb-4ffb-b732-7151d015b3b4.mp4

And better yet, the possibility to define the animation and background color? 🙂

artemiomorales commented:
> Has it been considered, instead of a zoom in motion, which in my opinion can be a bit too much, something like this?

I like this approach better. It's much more straightforward to implement and, as I'm looking at the code, I think it would be challenging to get the zoom to animate consistently across all WordPress sites.

Poliuk commented:
I also prefer Bea's animation 😄 

> the possibility to define the animation and background color? 🙂

That's a great question, and I think we haven't thought about it. I guess that if we add a setting to customize 
 the background and/or the animation, that setting should live inside some sort of advanced settings for each behavior, and they should be hidden by default. 

I would suggest leaving that for a second phase. Once we have the first version polished, we can start thinking about adding support for customization and adding some more animations.

jasmussen commented:
> That's a great question, and I think we haven't thought about it. I guess that if we add a setting to customize
the background and/or the animation, that setting should live inside some sort of advanced settings for each behavior, and they should be hidden by default.

One thing that might thread the needle is to use the same scrim color as the site background color, at least for the initial version. That takes care of dark or light themes at least.

artemiomorales committed:
Remove image translation animation and add fade instead

artemiomorales commented:
> I can't get this working for some reason. I see the "Lightbox" checkbox in the Advanced panel, but toggling it on does nothing on the frontend. Tried adding/removing links in the link control – no dice 🤔

@jameskoster Do you have the `block hydration experiments` plugin for the Interactivity API enabled by chance? If so, that may be overriding the customizations in this PR and preventing you from seeing the changes.

artemiomorales commented:
I pushed up a change to the animation so that it fades instead of zooming:

https://user-images.githubusercontent.com/5360536/233146228-38ecd18e-1aa3-44c7-b71d-4f6e2b7e79b4.mp4

Next will look at accessibility while discussion on the UI continues.

richtabor commented:
> I pushed up a change to the animation so that it fades instead of zooming

I prefer the zooming, though if the consensus is to fade that's fine. 🤷‍♂️

richtabor commented:
While testing today I noticed the front-end admin bar disappearing if the block/control is present: 

https://user-images.githubusercontent.com/1813435/233200530-c22855da-eb73-4631-aa8d-f2341b11c95b.mp4

michalczaplinski commented on packages/block-library/package.json:
```diff
@@ -27,7 +27,7 @@
 		"build-style/**",
 		"src/**/*.scss",
-		"{src,build,build-module}/*/init.js"
+		"{src,build,build-module}/**/init.js"
```
Just noting that this will expand the list of files that are allowed to have side effects so let's make sure that this change is intentional. I ran into the same thing in https://github.com/WordPress/gutenberg/pull/49305#discussion_r1151033211

artemiomorales replied:
This has been changed. Thanks! :D

luisherranz committed:
Add interactivity runtime

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add it to the image block

This is still pretty basic, just to check that it works.

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add a separate webpack config

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Make sure the runtime is imported only once

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Use sideEffects instead of init

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Move script registration to a general file

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Add `defer` to the interactivity scripts

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Revert changes of the image block

Co-authored-by: David Arenas <david.arenas@automattic.com>

luisherranz committed:
Fix init import name

Co-authored-by: David Arenas <david.arenas@automattic.com>

artemiomorales committed:
Add accessibility; clean up styles; fix bug regarding ref in directives

Cleaned up the PHP file a bit as I implemented standard accessibility
for a lightbox.

I changed names of attributes and properties to be clearer, as well as
conslidated some CSS.

Important to note: I fixed a bug with the directives wherein 'ref'
was coming in as null due to how it was being passed in hooks.js.

artemiomorales commented:
I just pushed a commit that adds accessibility. 

https://user-images.githubusercontent.com/5360536/233706484-9620317d-035e-49ef-aa10-e4e360505ea5.mp4

1. When opening the image using the keyboard, focus should go to the Close button.
2. Pressing escape or tabbing outside of the lightbox should close it.
3. Focus should revert the image.

I tested with a screen reader, and I think it provides ezough detail but I'm not sure — would appreciate a second pair of eyes on that 🙏

Next will work on the admin bar bug and the dynamic scrim color.

artemiomorales commented:
I forgot to mention that adding a link to the image should also now override the lightbox setting

mtias commented:
I prefer zooming as well, particularly for single images, seems less heavy handed

mtias commented:
Though I think we could drop the background overlay :) — in any case, this is a good example of how the kind of effect that users can choose should be extensible. I can install a plugin that provides a different lightbox effect or modifies the default one. I'd be curious to see if single image differes from gallery lightbox in that regard.

artemiomorales committed:
Configure image to use new Interactivity API runtime included in Gutenberg

artemiomorales committed:
Remove viewScript from image config

artemiomorales committed:
Add Portal directive to Interactivity API runtime

artemiomorales committed:
Set scrim to site background color

artemiomorales committed:
Remove extraneous image CSS declaration

artemiomorales commented:
> While testing today I noticed the front-end admin bar disappearing if the block/control is present

@richtabor The latest push should fix this issue. Thanks for catching this!

> One thing that might thread the needle is to use the same scrim color as the site background color, at least for the initial version. That takes care of dark or light themes at least.

@jasmussen The scrim now uses the site background color 👍 

I did run into the [following bug](https://github.com/WordPress/gutenberg/issues/49693) related to how we retrieve the CSS config; it would be good to get this resolved before shipping.

> I prefer zooming as well, particularly for single images, seems less heavy handed

It seems like we have some divided opinions on the zoom versus the fade 😄
From my point of view, the fade in / fade out is much easier to implement, and it's already done.

For the zoom effect, the following considerations make it harder:

- The image needs to animate smoothly from  its current position to the center of the screen
- The only performant way that I see to do this would be to use the `scale` and `translate` animatable CSS properties, as the `flexbox` and `margin` methods of centering won't work
- That means that we need to calculate the image's new properties based on the screen dimensions, then retrieve that from the HTML, which requires using directives to generate additional markup
- When the image is clicked, you can use its offset to calculate its would be new position, however that information becomes inaccurate as you change the scale, and I haven't been able to center it properly yet
- On top of this, it may be difficult to ensure the animation happens smoothly depending on the website layout, whether we're dealing with one column, two columns, or other, and mobile dimensions may be another special case to deal with

In all, there are a few unknowns with getting the zoom effect to work and I'm not sure how long they'd take to resolve, whereas the fade in / fade out already works, and is a well established paradigm that will play well with most, if not all, website layouts. So it's a question of priority and resources. I think the fade is good for now, and we can always circle back on adding the zoom later.

> Though I think we could drop the background overlay :) 

@mtias Can you clarify? Do you mean that there shouldn't be a default scrim color by default?

> in any case, this is a good example of how the kind of effect that users can choose should be extensible. I can install a plugin that provides a different lightbox effect or modifies the default one. I'd be curious to see if single image differs from gallery lightbox in that regard.

@mtias Agreed with this. Do you think we should aim to ship this feature with that extensibility?

Other than that, I think this PR is mostly done for now while discussion develops on [Outline for Block Behaviors](https://github.com/WordPress/gutenberg/issues/50029) and the [PR for the behaviors UI](https://github.com/WordPress/gutenberg/pull/49972) develops.

Another discussion we could have on the image lightbox specifically: Where should we put the local and global configuration for the scrim color? But maybe that's not a high priority for now, and perhaps it could be implicated in the global behaviors UI discussion.

artemiomorales commented:
Pinging @alexstine for accessibility feedback. Any guidance appreciated! 🙂

joedolson commented:
> Though I think we could drop the background overlay :)

For accessibility & usability, the background overlay provides affordance that elements behind the lightbox are inert, and can't be interacted with.

dmsnell commented on packages/block-library/src/image/edit.js:
```diff
@@ -238,10 +238,16 @@ export function ImageEdit( {
+		let enableLightbox = attributes.enableLightbox;
+		if ( ! enableLightbox ) {
+			enableLightbox = true;
+		}
```
if we don't need to modify this we can prefer a `const` in a single assignment

```js
const enableLightbox = attributes.enableLightbox || true;
```

we can also destructure the attributes if we know they aren't `null` (which we must here since we access its property)

```js
const { enableLightbox = true } = attributes;
```

dmsnell commented on packages/block-library/src/image/index.php:
```diff
@@ -5,6 +5,33 @@
+
+function processor_img_src_is_null( $processor ) {
+	return $processor->get_attribute( 'src' ) === null;
+}
```
I find it a bit distracting to jump around to this and the next function and realize they are doing something very basic. is there a need to extract these into separate functions? I think there's a good chance it would be clearer if we embedded these inline so that we could see in situ what is happening.

on that point, but more philosophically, I also want to be careful when passing around a Tag Processor instance. it's fine to do, but I think edges into the blurry territory of where things can get confused. it's nice when all of the modification happening to an HTML snippet/tag processor is happening in the same place vs. scattered across several places where we have to track them mentally.

just some thoughts on style.

artemiomorales replied:
Keeping everything inline and not passing around the tag processor instance sounds good to me 👍

dmsnell commented on packages/block-library/src/image/index.php:
```diff
@@ -5,6 +5,33 @@
+	}
+
+	return $processor;
+}
```
same note as above for `processor_img_src_is_null` - the function name is almost a direct copy of the function body, and if inlined, would remove a step of indirection

dmsnell commented on packages/block-library/src/image/index.php:
```diff
@@ -5,6 +5,33 @@
+		return false;
+	}
+	return false;
+}
```
small thing here, but it seems like this reduces down to a single return expression

```php
function has_link_destination( $attributes ) {
	return isset( $attributes['linkDestination'] ) && 'none' !== $attributes['linkDestination'];
}
```

dmsnell commented on packages/block-library/src/image/index.php:
```diff
@@ -14,29 +41,78 @@
+
+	$background_color = wp_get_global_styles(['color', 'background']);
+
+	if( ! has_link_destination( $attributes ) && isset( $attributes['enableLightbox'] ) && $attributes['enableLightbox'] === true ) {
```
seems a bit strange here to abstract `has_link_destination` but then use the inner logic from that inline with `enableLightbox`

I wonder if a different helper could be more generic and more reusable without coupling to the logic here.

```php
function get_or( $array, $key, $default ) {
	return array_key_exists( $key, $array )
		? $array[ $key ]
		: $default;
}

if (
	'none' === get_or( $attributes, 'linkDestination', 'none' ) &&
	true === get_or( $attributes, 'enableLightbox', false )
) {
	…
}
```

on the other hand, and given that we only seem to call `has_link_destination` here, we could also inline everything and it wouldn't be worse off, or give more names to what we're doing.

```php
$link_destination = isset( $attributes['linkDestination'] ) ? $attributes['linkDestination'] : 'none';
$enable_lightbox  = isset( $attributes['enableLightbox'] ) ? $attributes['enableLightbox'] : false;

if ( 'none' === $link_destination && $enable_lightbox ) {
	…
}
```

artemiomorales replied:
Second approach looks good and clean to me 👍

dmsnell commented on packages/block-library/src/image/index.php:
```diff
@@ -14,29 +41,78 @@
+		$modal_content = new WP_HTML_Tag_Processor( $content );
+		$modal_content->next_tag( 'img' );
+		$modal_content = process_img_data_id($modal_content, $attributes);
+		$modal_content = $modal_content->get_updated_html();
```
what's supposed to be different here between `$body_content` and `$modal_content`?

it seems like they will be the same, except if the `IMG` `src` is missing then `$body_content` will be empty but the modal won't be.

artemiomorales replied:
This is a distinction left over from an earlier commit. Will remove the duplication!

dmsnell commented on packages/block-library/src/image/index.php:
```diff
@@ -14,29 +41,78 @@
+			return '';
+		}
+
+		$body_content = process_img_data_id($body_content, $attributes);
```
this is a good place to serve as an example of where I get a bit anxious seeing the processor passed around. we're setting `$body_content` to the return, which is the tag processor, but it was _already still the processor_ without the assignment. being stateful, it does stuff while hidden inside of `process_img_data_id`, and it's not obvious from reading this code that the `$body_content` Tag Processor in line 57 could be "different" from the one in line 56

dmsnell commented on packages/block-library/src/image/index.php:
```diff
@@ -14,29 +41,78 @@
+						 data-wp-class.initialized="context.core.initialized"
+						 data-wp-class.active="context.core.lightboxEnabled"
+						 data-wp-init.hideLightboxOnEsc='actions.core.hideLightboxOnEsc'
+						 data-wp-init.hideLightboxOnTab='actions.core.hideLightboxOnTab'
```
it doesn't matter from an HTML perspective if these are single or double quotes, but it could help in the example code if they are double like the others.

dmsnell commented on packages/block-library/src/image/index.php:
```diff
@@ -14,29 +41,78 @@
+							<div class="scrim" style='background-color: $background_color'></div>
+					</div>
+			</div>
+		HTML;
```
prior to PHP 7.3.0 the `HTML;` needed to start at the first character on the line. given that WordPress support goes back much further we may want to remove the leading indentation.

dmsnell commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -0,0 +1,101 @@
+				};
+			},
+			hideLightboxOnTab: ( { context } ) => {
+				async function handleTab( event ) {
```
nitpick: Is there a need for `async` here? it doesn't seem so, so we can probably remove it

artemiomorales replied:
Thanks for catching! This one is also left over from earlier ideas.

dmsnell commented on packages/block-library/src/utils/interactivity/directives.js:
```diff
@@ -51,6 +53,28 @@ export default () => {
+						<Provider value={inheritedValue}>{children}</Provider>,
+						document.body
+					) }
+				</>
```
it seems like we could eliminate the fragment `<>` here since `createPortal()` isn't JSX anyway, unless I'm missing something and it somehow returns a list of things that's not a list?

```js
return createPortal(
	<Provider value={inheritedValue}>{children}</Provider>,
	document.body
);
```

dmsnell commented on packages/block-library/src/utils/interactivity/portals.js:
```diff
@@ -0,0 +1,82 @@
+	const _this = this;
+	let container = props._container;
+
+	_this.componentWillUnmount = function () {
```
if we use arrow-functions we shouldn't need to worry about `this` assignment, which I'm guessing is why we set `_this = this`?

dmsnell replied:
on the other hand we might be able to use the `class` form and not worry about it either.

artemiomorales replied:
This file is copied and pasted from the [portals.js](https://github.com/preactjs/preact/blob/master/compat/src/portals.js) file in the Preact library.

Do you think we should introduce changes to this file or keep it in sync with Preact?

dmsnell replied:
I doubt this file will stay in sync with `preact`. if that's the goal we might want to directly import it instead of importing our copy of it.

not a big deal, but I haven't seen `_this` since around late 2014, and I think folks were happy to give it up. can leave it as-is for now, or explore a way to import/require the direct file from `preact`

dmsnell commented on packages/block-library/src/utils/interactivity/portals.js:
```diff
@@ -0,0 +1,82 @@
+					this.childNodes.splice(this.childNodes.indexOf(child) >>> 1, 1);
+					_this._container.removeChild(child);
+				}
+			};
```
what's the purpose of the fake nodes? why not create actual DOM nodes here?

dmsnell commented on packages/block-library/src/utils/interactivity/portals.js:
```diff
@@ -0,0 +1,82 @@
+					_this._container.appendChild(child);
+				},
+				removeChild(child) {
+					this.childNodes.splice(this.childNodes.indexOf(child) >>> 1, 1);
```
what's the purpose of the `>>> 1` here? does this not remove a node approximately half-way to the requested child? or try to remove the 2147483647th item from the list if the child isn't part of the `childNodes`?

dmsnell commented on packages/block-library/src/utils/interactivity/portals.js:
```diff
@@ -0,0 +1,82 @@
+	// When we come from a conditional render, on a mounted
+	// portal we should clear the DOM.
+	else if (_this._temp) {
+		_this.componentWillUnmount();
```
it seems a bit suspicious to call `componentWillUnmount()` manually. maybe we could create a method with a name that communicates what it's doing to unload or un-render, and then inside `componentWillUnmount()` we call _that_ function? then we wouldn't have to cross abstraction boundaries by calling the lifecycle methods when we _want_ to do something else (in this case, do the thing that happens on unmount)

dmsnell reviewed (commented):
I know there's a lot of higher-level discussion going on here but I wanted to drop by and share some technical-level thoughts and review. happy to chat more about them if you want.

afercia commented:
Agree with @joedolson that the modal dialog component should cover most of the accessibility requirements. Still:
- focus indication needs to be always visible
- ideally, the modal dialog should have a visible Close button
- the trigger that opens the modal dialog must be a `<button>` element
- the button trigger must be properly labeled to tell users what it does
- the button trigger must be visible and its accessible name must be exposed in some way (e.g. for speech recognition software users)

> It's very likely that the link that triggers the lightbox doesn't need to be visible until someone uses the assistive technologies. In effect, it could use the screen-reader-text CSS class that exists in WordPress core.

I'm not sure that would be okay for sighted keyboard users.

joedolson commented:
> It's very likely that the link that triggers the lightbox doesn't need to be visible until someone uses the assistive technologies. In effect, it could use the screen-reader-text CSS class that exists in WordPress core.

I'm not sure what you mean by "uses the assistive technologies". What you're describing sounds like the same thing we do with skip links, exposing them on focus, which is not dependent on assistive technology; just focus. We should not (and, for all practical purposes, cannot) detect the use of assistive technology.

gziolo committed:
Move and refactor the interactive scritps registration

gziolo committed:
Fix code style violations

gziolo committed:
Use `wp-interactivity-` prefix for script handles

gziolo committed:
Improve the matcher for side effects in `package.json`

artemiomorales committed:
Improve aria labeling

artemiomorales commented:
Thanks for the feedback @afercia! Here are my responses:

> Agree with @joedolson that the modal dialog component should cover most of the accessibility requirements. Still:

To speak to this, since this PR is about using the new Interactivity API, we're unable to build on top of the existing implementation, as the markup is defined in PHP and the interactivity is wired through the API rather than React components. That being said, I've looked at the modals for Navigation as well as the Modal component for reference while doing this.

One thing I've omitted is a tooltip, but I'm not not sure if it makes sense to use one here.

> 1. focus indication needs to be always visible

Can you clarify what you mean here? Currently, you can tab to an image with a lightbox, which receives focus and when you press Enter, it opens the modal and sets focus to the close button in the top right corner.

<img width="1464" alt="Screen Shot 2023-04-28 at 10 32 45 AM" src="https://user-images.githubusercontent.com/5360536/235626811-98a33d88-6d64-42f3-af8f-a8c56d784446.png">

> 2. ideally, the modal dialog should have a visible Close button

Currently, the modal has a close button displayed as an X icon. Do you mean that it should also be labeled with the word, 'Close?'

> 3. the trigger that opens the modal dialog must be a <button> elemenat
> 4. the button trigger must be properly labeled to tell users what it does

This is indeed how the modal is triggered, and I believe it's properly labeled. I just pushed up an improvement that makes the label clearer than it was before — it now reads 'Open image lightbox' along with the image's alt text, if available.

> 5. the button trigger must be visible and its accessible name must be exposed in some way (e.g. for speech recognition software users)

Do you mean that the button label should be visible and readable in the UI for all users?

dmsnell commented on lib/client-assets.php:
```diff
@@ -574,3 +574,50 @@ function gutenberg_register_vendor_scripts( $scripts ) {
+ * admin panel.
+ */
+function gutenberg_register_interactivity_scripts() {
+ 	if ( ! is_admin() )	 {
```
quick note here and in `add_defer_attribute()` - another technical nit-pick

we generally prefer early-abort over nesting large code bodies, so we could rearrange these by inverting the conditions

```php
function gutenberg_register_interactivity_scripts() {
	if ( is_admin() ) {
		return;
	}

	// rest of the function
}

function add_defer_attribute( $tag, $handle ) {
	if ( ! str_starts_with( $handle, 'interactivity-' ) ) {
		return $tag;
	}

	$p = new WP_HTML_Tag_Processor( $tag );
	// rest of the code
	return $p->get_updated_html();
}
```

the value for the preference is being able to know immediately when we exit the function, not needing to continue holding on to the conditional in case the function does something after the end of the indented body

artemiomorales committed:
Code cleanup; simplify syntax, consolidate code

artemiomorales committed:
Refactor code, remove event listeners, consolidate logic

artemiomorales committed:
Fix formatting in SCSS file

artemiomorales committed:
Change CheckboxControl to a ToggleControl; update API docs

artemiomorales committed:
Update wp_enqueue_script to correctly add interactivity runtime

artemiomorales committed:
Fix linter errors

DAreRodz committed:
Add custom useSignalEffect

artemiomorales committed:
Update to use core.image namespace

artemiomorales committed:
Pause closing of lightbox slightly when using the mousewheel

artemiomorales committed:
Rename portal directive to 'wp-body' and remove unused reference

artemiomorales committed:
Add internal dependencies flag; update comment

artemiomorales committed:
Remove extraneous code

artemiomorales commented:
I pushed cleanups to the code incorporating @dmsnell’s feedback. I’ve also been discussing improvements to the code with @SantosGuillamot and @luisherranz. Here follows a summary of our discussion.

## Event Listeners

Mario suggested that we declare the event listeners on the lightbox upfront rather than attaching and detaching them whenever it is opened based on his work on the [navigation block migration](https://github.com/WordPress/gutenberg/pull/50041). The resulting code is much cleaner and his been implemented.

Luis initially expressed some reservations with this approach because it would mean attaching many event listeners to the DOM; however, the events are only fired when the modal is open, so it seems like this approach won’t cause much overhead.

## Markup Size Explorations

Another point of discussion has been that right now, the modal markup is delivered for each image, and we considered whether it would be better to limit the amount of markup by either 1.) using an HTML singleton for the lightbox that is populated whenever it is opened, or 2.) using Preact state to remove the markup from the DOM. The purpose here would be to reduce the size of the HTML delivered to end users.

However, while doing explorations, we discovered that the size of the lightbox markup without the containing image is just 457 bytes when gzipped. In addition, additional instances of the lightbox markup only increase the delivered size by a miniscule amount — 10 instances of it would amount to just 503 bytes.

The image markup is also duplicated within the lightbox, but in my test of a figure element of 200 bytes, the duplicated instances only amounted to an additional 50 bytes per image.

**So in sum, we’re looking at a footprint of about 500 bytes for the first lightbox, then about 55 bytes for each lightbox thereafter.**

With this in mind, I don’t think it makes sense to find a way to deliver the lightbox markup as a singleton and introduce more complicated logic when the footprint is so small, seemingly a micro optimization that is less important than a good developer experience.

As for removing the markup from the DOM, this doesn’t actually reduce the size of the HTML that is delivered to the client — it just hides it from view. I also found that when doing so, we lose the CSS fade out animation because Preact removes the item from the DOM before the animation can take place. Finding a way to make this logic work nicely with the CSS animation would introduce unnecessary complexity for little gain.


## Conclusion

With all of this in mind, the PR currently includes the consolidation of the event listeners but does not address the duplication of the markup in the DOM, so we’ll see the markup for each lightbox at the level of the body.

I think this approach makes for a good developer experience while still having minimal impact on the overall size of the HTML and is the best approach.

That being said, I’m happy to hear other opinions or be proven wrong. Let me know your thoughts. Thanks!

DAreRodz committed:
Call `init` after `store` has been initialized

SantosGuillamot commented on packages/block-library/src/image/interactivity.js:
```diff
@@ -0,0 +1,81 @@
+ */
+import { store } from '../utils/interactivity';
+
+const raf = window.requestAnimationFrame;
+// Until useSignalEffects is fixed: https://github.com/preactjs/signals/issues/228
+const tick = () => new Promise( ( r ) => raf( () => raf( r ) ) );
```
Now that [this pull request](https://github.com/WordPress/block-interactivity-experiments/pull/226) has been merged, this workaround shouldn't be necessary. Could you please try merging the `add-interactivity-runtime` branch and check if everything works as expected without using `await tick()`?

SantosGuillamot commented on packages/block-library/src/image/edit.js:
```diff
@@ -111,6 +111,7 @@ export function ImageEdit( {
 		width,
 		height,
 		sizeSlug,
+		enableLightbox = true,
```
Not 100% sure but, shouldn't we define the [default value](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/#default-value) in the `block.json` instead of here? Something like:

```json
"enableLightbox": {
    "type": "boolean",
    "default": true
}
```

artemiomorales commented:
Looks like this was closed when the base branch was merged; reopening now.

gziolo commented:
@artemiomorales, maybe try to rebase with `trunk` in your fork. There is no option to reopen PR at the moment. However, I see that the branch is still there 🤷🏻

artemiomorales commented:
Hi all, it appears we'll be unable to recuperate this pull request, so I've [opened a new one here](https://github.com/WordPress/gutenberg/pull/50373). Let's continue the conversation there. Thanks!

artemiomorales commented:
Hi all, here I'm giving a call for feedback as I've just pushed a batch of changes to the [new pull request](https://github.com/WordPress/gutenberg/pull/50373). The admin UI will change slightly when it's synced with [WIP: UI Behavior](https://github.com/WordPress/gutenberg/pull/49972), but other than that, I believe this PR is done.

Would be great to hear feedback on any of the decisions or revisions made 😄
